opm/ml/ml_model.hpp

```diff
@@ -0,0 +1,437 @@
+                                                        "{}"
+                                                        "  max: "
+                                                        "{}",
+                                                        k, dims_[2]));
+    OPM_ERROR_IF(!(l < dims_[3] && l >= 0), fmt::format(" Invalid l:  "
+                                                        "{}"
+                                                        "  max: "
+                                                        "{}",
+                                                        l, dims_[3]));
+
+    return data_[dims_[3] * (dims_[2] * (dims_[1] * i + j) + k) + l];
+  }
+  void fill(const T& value) { std::fill(data_.begin(), data_.end(), value); }
+
+  // Tensor addition
+  Tensor operator+(const Tensor& other) {
+    OPM_ERROR_IF(dims_.size() != other.dims_.size(),
+                 "Cannot add tensors with different dimensions");
+    Tensor result;
+    result.dims_ = dims_;
+    result.data_.reserve(data_.size());
+
+    std::transform(data_.begin(), data_.end(), other.data_.begin(),
+                   std::back_inserter(result.data_),
+                   [](T x, T y) { return x + y; });
+
+    return result;
+  }
+
+  // Tensor multiplication
+  Tensor multiply(const Tensor& other) {
+    OPM_ERROR_IF(dims_.size() != other.dims_.size(),
+                 "Cannot multiply elements with different dimensions");
+
+    Tensor result;
+    result.dims_ = dims_;
+    result.data_.reserve(data_.size());
+
+    std::transform(data_.begin(), data_.end(), other.data_.begin(),
+                   std::back_inserter(result.data_),
+                   [](T x, T y) { return x * y; });
+
+    return result;
+  }
+
+  // Tensor dot for 2d tensor
+  Tensor dot(const Tensor& other) {
+    OPM_ERROR_IF(dims_.size() != 2, "Invalid tensor dimensions");
+    OPM_ERROR_IF(other.dims_.size() != 2, "Invalid tensor dimensions");
+
+    OPM_ERROR_IF(dims_[1] != other.dims_[0],
+                 "Cannot multiply with different inner dimensions");
+
+    Tensor tmp(dims_[0], other.dims_[1]);
+
+    for (int i = 0; i < dims_[0]; i++) {
+      for (int j = 0; j < other.dims_[1]; j++) {
+        for (int k = 0; k < dims_[1]; k++) {
+          tmp(i, j) += (*this)(i, k) * other(k, j);
+        }
+      }
+    }
+
+    return tmp;
+  }
+
+  std::vector<int> dims_;
+  std::vector<T> data_;
+};
+
+// NN layer
+// ---------------------
+/** \class Neural Network  Layer base class
+ */
+template <class Evaluation> class NNLayer {
+ public:
+  NNLayer() {}
+
+  virtual ~NNLayer() {}
+
+  virtual bool loadLayer(std::ifstream& file) = 0;
+
+  virtual bool apply(const Tensor<Evaluation>& in, Tensor<Evaluation>& out) = 0;
+};
+
+/** \class Activation  Layer class
+ * Applies an activation function
+ */
+template <class Evaluation>
+class NNLayerActivation : public NNLayer<Evaluation> {
+ public:
+  enum ActivationType {
```

Comment:
It was requested to make this an enum class, and you commented that this was done, but it is not done.